sql/group.go

```diff
@@ -0,0 +1,439 @@
+	// TODO(pmattis): This only handles aggregate functions, not GROUP BY.
+
+	// Loop over the render expressions and extract any aggregate functions.
+	var funcs []*parser.FuncExpr
+	for _, r := range s.render {
+		f, err := extractAggregateFuncs(r)
+		if err != nil {
+			return nil, err
+		}
+		funcs = append(funcs, f...)
+	}
+	if len(funcs) == 0 {
+		return nil, nil
+	}
+
+	// Aggregation is being performed. Loop over the render expressions again and
+	// verify that the only qvalues mentioned by GROUP BY expressions are allowed
+	// outside of the aggregate function arguments. For example, the following is
+	// illegal because k is used outside of the aggregate function and not part
+	// of a GROUP BY expression.
+	//
+	//   SELECT COUNT(k), k FROM kv GROUP BY v
+	for _, r := range s.render {
+		if err := checkAggregateExpr(r); err != nil {
+			return nil, err
+		}
+	}
+
+	if log.V(2) {
+		log.Infof("Group: %s", funcs)
+	}
+
+	group := &groupNode{
+		columns: s.columns,
+		render:  s.render,
+		funcs:   funcs,
+		aggs:    make([]aggregateFunc, 0, len(funcs)),
+	}
+
+	for _, f := range funcs {
+		agg, ok := getAggregateFunc(f)
+		if !ok {
+			panic(fmt.Sprintf("%s not found (this shouldn't happen)", f.Name))
+		}
+		if agg == nil {
+			return nil, fmt.Errorf("%s not implemented", f.Name)
+		}
+		group.aggs = append(group.aggs, agg.New())
+	}
+
+	// Replace the render expressions in the scanNode with expressions that
+	// compute only the arguments to the aggregate expressions.
+	s.columns = make([]string, 0, len(funcs))
+	s.render = make([]parser.Expr, 0, len(funcs))
+	for _, f := range funcs {
+		if len(f.Exprs) != 1 {
+			panic(fmt.Sprintf("%s has %d arguments (expected 1)", f.Name, len(f.Exprs)))
+		}
+		s.columns = append(s.columns, f.String())
+		s.render = append(s.render, f.Exprs[0])
+	}
+
+	return group, nil
+}
+
+type groupNode struct {
+	plan      planNode
+	columns   []string
+	row       parser.DTuple
+	render    []parser.Expr
+	funcs     []*parser.FuncExpr
+	aggs      []aggregateFunc
+	needGroup bool
+	err       error
+}
+
+func (n *groupNode) Columns() []string {
+	return n.columns
+}
+
+func (n *groupNode) Ordering() []int {
+	return n.plan.Ordering()
+}
+
+func (n *groupNode) Values() parser.DTuple {
+	return n.row
+}
+
+func (n *groupNode) Next() bool {
+	if !n.needGroup {
+		return false
+	}
+	n.needGroup = false
+
+	// Loop over the rows passing the values into the corresponding aggregation
+	// functions.
+	for n.plan.Next() {
+		values := n.plan.Values()
+		for i, a := range n.aggs {
+			if n.err = a.Add(values[i]); n.err != nil {
+				return false
+			}
+		}
+	}
+
+	n.err = n.plan.Err()
+	if n.err != nil {
+		return false
+	}
+
+	// Replace the aggregate function argument with the result of the aggregate
+	// function. Really this should be replacing the entire function expression,
+	// but this is easier. Note that the aggregate function exprssion evaluation
```

Comment:
this is easier because it allows you to keep the type checking in the dispatch code? i'm not really clear on the interaction between this, the type assertions in the aggregates themselves, and the type checking in dispatch.